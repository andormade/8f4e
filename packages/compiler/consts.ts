export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
