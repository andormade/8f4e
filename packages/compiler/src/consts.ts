export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
export const LOGIC_HIGH = I16_SIGNED_LARGEST_NUMBER;
export const LOGIC_LOW = 0;
