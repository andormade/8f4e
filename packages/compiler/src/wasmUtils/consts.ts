export const WASM_MEMORY_PAGE_SIZE = 65536;
