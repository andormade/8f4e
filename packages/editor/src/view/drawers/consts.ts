export const VGRID = 8;
export const HGRID = 16;
