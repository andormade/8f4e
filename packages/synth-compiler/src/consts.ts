export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I32_SIGNED_LARGEST_NUMBER = 2147483647;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
export const LOGIC_HIGH = I16_SIGNED_LARGEST_NUMBER;
export const LOGIC_LOW = 0;
export const WORD_LENGTH = 4;
