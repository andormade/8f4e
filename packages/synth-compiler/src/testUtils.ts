import {
	createFunctionSection,
	createTypeSection,
	createFunctionType,
	createExportSection,
	createFunctionExport,
	createCodeSection,
	createImportSection,
	createMemoryImport,
	FunctionBody,
} from '@8f4e/bytecode-utils';
import { compile } from '@8f4e/module-compiler';
import { getInitialMemory, ModuleGenerator } from '.';
import { CompiledModule, MemoryMap } from './types';
import wabt from 'wabt';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: Int32Array, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		memory[i] = initialMemory[i];
	}
}

export async function createTestModule(
	moduleCreator: ModuleGenerator | string,
	initialConfig = {}
): Promise<{
	memory: Int32Array;
	test: CallableFunction;
	reset: () => void;
	wat: string;
	program: Uint8Array;
	memoryMap: MemoryMap[];
}> {
	let module: CompiledModule;

	if (typeof moduleCreator === 'function') {
		module = moduleCreator('test', { byte: nthWord => nthWord * 4, word: nthWord => nthWord }, initialConfig);
	} else {
		module = compile(moduleCreator, 'test', 0);
	}

	const program = createSingleFunctionWASMProgram(module.functionBody);

	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const memoryBuffer = new Int32Array(memoryRef.buffer);

	const { instance } = await WebAssembly.instantiate(program, {
		js: {
			memory: memoryRef,
		},
	});

	const reset = () => {
		setInitialMemory(memoryBuffer, module);
	};

	reset();

	const test = instance.exports.test as CallableFunction;

	const wat: string = await new Promise(resolve => {
		wabt().then(_wabt => {
			const module = _wabt.readWasm(program, {});
			resolve(module.toText({}));
		});
	});

	return { memory: memoryBuffer, test, reset, wat, program, memoryMap: module.memoryMap };
}
