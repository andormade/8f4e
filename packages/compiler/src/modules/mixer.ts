import { i32load, i32const, i32store, localSet, localGet, ifelse } from '../wasm/instructions';
import { createFunctionBody, createLocalDeclaration } from '../wasm/sections';
import { Instruction, Type } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER, I16_SIGNED_SMALLEST_NUMBER } from '../consts';

export const enum Memory {
	DEFAULT_VALUE,
	OUTPUT,
	INPUT_POINTER_1,
	INPUT_POINTER_2,
	INPUT_POINTER_3,
	INPUT_POINTER_4,
}

const enum Locals {
	RESULT,
	__LENGTH,
}

const mixer: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			...i32const(offset(Memory.OUTPUT)),

			...[
				...i32const(offset(Memory.INPUT_POINTER_1)),
				...i32load(),
				...i32load(),

				...i32const(offset(Memory.INPUT_POINTER_2)),
				...i32load(),
				...i32load(),

				Instruction.I32_ADD,

				...i32const(offset(Memory.INPUT_POINTER_3)),
				...i32load(),
				...i32load(),

				Instruction.I32_ADD,

				...i32const(offset(Memory.INPUT_POINTER_4)),
				...i32load(),
				...i32load(),

				Instruction.I32_ADD,
				...localSet(Locals.RESULT),

				// Limit
				...localGet(Locals.RESULT),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				Instruction.I32_GE_S,
				...ifelse(Type.I32, [...i32const(I16_SIGNED_LARGEST_NUMBER)], [...localGet(Locals.RESULT)]),
				...localSet(Locals.RESULT),

				...localGet(Locals.RESULT),
				...i32const(I16_SIGNED_SMALLEST_NUMBER),
				Instruction.I32_LE_S,
				...ifelse(Type.I32, [...i32const(I16_SIGNED_SMALLEST_NUMBER)], [...localGet(Locals.RESULT)]),
			],

			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset: offset(0),
		initialMemory: [
			0,
			0,
			offset(Memory.DEFAULT_VALUE),
			offset(Memory.DEFAULT_VALUE),
			offset(Memory.DEFAULT_VALUE),
			offset(Memory.DEFAULT_VALUE),
		],
		memoryAddresses: [
			{ address: offset(Memory.INPUT_POINTER_1), id: 'in:1' },
			{ address: offset(Memory.INPUT_POINTER_2), id: 'in:2' },
			{ address: offset(Memory.INPUT_POINTER_3), id: 'in:3' },
			{ address: offset(Memory.INPUT_POINTER_4), id: 'in:4' },
			{ address: offset(Memory.OUTPUT), id: 'out' },
		],
	};
};

export default mixer;
