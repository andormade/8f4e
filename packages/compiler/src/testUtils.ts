import wabt from 'wabt';

import {
	createCodeSection,
	createExportSection,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import { FunctionBody } from './wasmUtils/typeHelpers';
import { CompiledModule, TestModule } from './types';
import { WORD_LENGTH } from './consts';

import { compileModules, getInitialMemory } from '.';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: DataView, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		if (Number.isInteger(initialMemory[i])) {
			memory.setInt32(i * 4, initialMemory[i], true);
		} else {
			memory.setFloat32(i * 4, initialMemory[i], true);
		}
	}
}

export async function createTestModule(sourceCode: string): Promise<TestModule> {
	let allocatedMemoryForTestData = 0;

	const module: CompiledModule = compileModules([{ code: sourceCode.split('\n') }])[0];
	const program = createSingleFunctionWASMProgram(module.functionBody);
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const dataView = new DataView(memoryRef.buffer);
	const memoryBuffer = new Int32Array(memoryRef.buffer);
	let instance: WebAssembly.Instance | undefined;
	let wat = '';

	try {
		const webAssemblyInstantiatedSource = await WebAssembly.instantiate(program, {
			js: {
				memory: memoryRef,
			},
		});
		instance = webAssemblyInstantiatedSource.instance;

		wat = await new Promise(resolve => {
			wabt().then(_wabt => {
				const module = _wabt.readWasm(program, {});
				resolve(module.toText({}));
			});
		});
	} catch (error) {
		console.log(error);
	}

	const reset = () => {
		setInitialMemory(dataView, module);

		// Clear the test data that was out of the module's scope.
		for (let i = module.memoryWordSize; i < module.memoryWordSize + allocatedMemoryForTestData; i++) {
			memoryBuffer[i] = 0;
		}
		allocatedMemoryForTestData = 0;
	};

	reset();

	const test = (instance?.exports.test ||
		function () {
			return;
		}) as CallableFunction;

	const memoryGet = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);

			if (!memoryItem) {
				return;
			}

			if (memoryItem.isInteger) {
				return dataView.getInt32(memoryItem.byteAddress, true);
			} else {
				return dataView.getFloat32(memoryItem.byteAddress, true);
			}
		}

		return memoryBuffer[address];
	};

	const memorySet = (address: string | number, value: number | number[]): void => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			if (typeof value === 'number') {
				if (Number.isInteger(value)) {
					dataView.setInt32(memoryItem.byteAddress, value, true);
				} else {
					dataView.setFloat32(memoryItem.byteAddress, value, true);
				}
			} else {
				for (let i = 0; i < value.length; i++) {
					if (Number.isInteger(value[i])) {
						dataView.setInt32(memoryItem.byteAddress + i * 4, value[i], true);
					} else {
						dataView.setFloat32(memoryItem.byteAddress + i * 4, value[i], true);
					}
				}
			}
			return;
		}

		if (typeof value === 'number') {
			if (Number.isInteger(value)) {
				dataView.setInt32(address * 4, value, true);
			} else {
				dataView.setFloat32(address * 4, value, true);
			}
		} else {
			for (let i = 0; i < value.length; i++) {
				if (Number.isInteger(value[i])) {
					dataView.setInt32(address * 4 + i * 4, value[i], true);
				} else {
					dataView.setFloat32(address * 4 + i * 4, value[i], true);
				}
			}
		}
	};

	const getByteAddress = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			return memoryItem.byteAddress;
		}

		return address * WORD_LENGTH;
	};

	const allocMemoryForPointer = (address: string | number): number => {
		const firstFreeAddress = module.memoryWordSize + allocatedMemoryForTestData;
		memorySet(address, firstFreeAddress * WORD_LENGTH);
		allocatedMemoryForTestData++;
		return firstFreeAddress;
	};

	// @ts-ignore
	memoryBuffer.get = memoryGet;
	// @ts-ignore
	memoryBuffer.set = memorySet;
	// @ts-ignore
	memoryBuffer.allocMemoryForPointer = allocMemoryForPointer;
	// @ts-ignore
	memoryBuffer.byteAddress = getByteAddress;

	return {
		// @ts-ignore
		memory: memoryBuffer,
		test,
		reset,
		wat,
		program,
		memoryMap: module.memoryMap,
		ast: module.ast,
	};
}
