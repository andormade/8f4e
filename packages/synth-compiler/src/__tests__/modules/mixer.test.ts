import { createTestModule } from '../../testUtils';
import mixer, { Memory } from '../../modules/mixer';
import { I16_SIGNED_LARGEST_NUMBER, I16_SIGNED_SMALLEST_NUMBER } from '../../consts';

let testModule;

describe('funciontal tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(mixer);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test('negate module', () => {
		const { memory, test } = testModule;

		memory[Memory.INPUT_POINTER_1] = 10 * memory.BYTES_PER_ELEMENT;
		memory[Memory.INPUT_POINTER_2] = 11 * memory.BYTES_PER_ELEMENT;
		memory[Memory.INPUT_POINTER_3] = 12 * memory.BYTES_PER_ELEMENT;
		memory[Memory.INPUT_POINTER_4] = 13 * memory.BYTES_PER_ELEMENT;

		memory[10] = 1;
		memory[11] = 1;
		memory[12] = 1;
		memory[13] = 1;
		test();
		expect(memory[Memory.OUTPUT]).toBe(4);

		memory[10] = 1;
		memory[11] = -1;
		memory[12] = 1;
		memory[13] = -1;
		test();
		expect(memory[Memory.OUTPUT]).toBe(0);

		memory[10] = I16_SIGNED_LARGEST_NUMBER;
		memory[11] = I16_SIGNED_LARGEST_NUMBER;
		memory[12] = I16_SIGNED_LARGEST_NUMBER;
		memory[13] = I16_SIGNED_LARGEST_NUMBER;
		test();
		expect(memory[Memory.OUTPUT]).toBe(I16_SIGNED_LARGEST_NUMBER);

		memory[10] = I16_SIGNED_SMALLEST_NUMBER;
		memory[11] = I16_SIGNED_SMALLEST_NUMBER;
		memory[12] = I16_SIGNED_SMALLEST_NUMBER;
		memory[13] = I16_SIGNED_SMALLEST_NUMBER;
		test();
		expect(memory[Memory.OUTPUT]).toBe(I16_SIGNED_SMALLEST_NUMBER);
	});
});
