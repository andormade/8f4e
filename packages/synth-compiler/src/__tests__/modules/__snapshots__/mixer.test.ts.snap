// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`funciontal tests if the generated memory map matches with the snapshot 1`] = `
Array [
  Object {
    "address": 0,
    "default": 0,
    "type": 5,
  },
  Object {
    "address": 1,
    "default": 0,
    "id": "out",
    "type": 4,
  },
  Object {
    "address": 2,
    "default": 0,
    "id": "in:1",
    "type": 3,
  },
  Object {
    "address": 3,
    "default": 0,
    "id": "in:2",
    "type": 3,
  },
  Object {
    "address": 4,
    "default": 0,
    "id": "in:3",
    "type": 3,
  },
  Object {
    "address": 5,
    "default": 0,
    "id": "in:4",
    "type": 3,
  },
]
`;

exports[`funciontal tests if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32)
    i32.const 4
    i32.const 8
    i32.load
    i32.load
    i32.const 12
    i32.load
    i32.load
    i32.add
    i32.const 16
    i32.load
    i32.load
    i32.add
    i32.const 20
    i32.load
    i32.load
    i32.add
    local.set 0
    local.get 0
    i32.const 32767
    i32.ge_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      local.get 0
    end
    local.set 0
    local.get 0
    i32.const -32768
    i32.le_s
    if (result i32)  ;; label = @1
      i32.const -32768
    else
      local.get 0
    end
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
